table/tables/mutation_checker.go

```diff
@@ -381,9 +381,31 @@ func getOrBuildColumnMaps(
 	return maps
 }
 
-// CheckTxnConsistency checks the number of row/index mutations before the txn commits,
+// CheckIndexCount checks the number of row/index mutations before the txn commits,
 // to prevent some inconsistent transactions.
-func CheckTxnConsistency(txn kv.Transaction) error {
+//
+// Claim: For mutations in each table in the transaction, #(row insertions) != 0 => #(index insertions) % #(row insertions) = 0
+// Proof:
+// Without loss of generality, assume there is only 1 table involved in the transaction.
+// We examine the changes of numbers of these mutations: PI(put index), PR(put record), DI(delete index), DR(delete record),
+//  for each single-row operations. We argue that delta(PI) / delta(PR) = N always holds, where N is the number of indices.
+// Let h denote the handle, v denote the value of the row (or corresponding index mutations).
+// add record(h, v)
+//    h does not exist
+//        v no duplicates (index does not exist): PR + 1, PI + N
+//        v duplicated (0 puts, M dels): PR + 1, DI - M, PI + N
+// 			note: duplicated index mutations can only come from deletions on unique indices
+//    h is del
```

Comment:
Does `h does not exist` and `h is del` indicates whether it exists in the current transaction's mem buffer?

Reply:
Yep. `h is del` means it exists in the membuf and is a del.